### PR TITLE
add ability to upload buffer

### PR DESCRIPTION
**Pull request by cpaczek**
### What does it do?

Ability to upload a nodejs file buffer. 

### Why is it needed?

Currently, the upload plugin wants a FormData object only available on the front end. this adds the ability to upload a node js buffer to the media library

### How to test it?

```js        
const file = fs.readFileSync('./test.png')
return strapi.plugin('upload').service('upload').uploadBuffer({file, fileName: crypto.randomUUID(), mime: 'image/png', ext: 'png', alternativeText: "test"})
```

### CODE REVIEW

**Comment by Cpaczek** [comment]
@joshuaellis

**Comment by Cpaczek** [comment]
Just re-read the original issue. seems like they want to be able to do this via the content API, this PR just allows users to do it using the Strapi object. This would make a custom controller a lot easier to implement though. 

If they wanted to use base64 they could convert to a buffer using the nodejS buffer class.
```
const base64Data = '...'; // Your base64 encoded data
const binaryData = Buffer.from(base64Data, 'base64');
```

**Comment by joshuaellis** [comment]
Hey @Cpaczek thanks for the PR! as this is a feature addition theres a bit of a stricter procedure on what we require for a PR to be merged.

Looking at what you've done so far:
* it'd be great to add tests for the service function you've added
* Can we utilise this service in the admin API and/or the content API controllers?

I'm also gonna request a review other engineers too 💪🏼

**Comment by Cpaczek** [comment]
@joshuaellis 
Ya, I can add some tests. In regard to adding it to content API it's not really designed to interact with REST as it uses a nodejs buffer, we could implement a solution that uses base 64 but the issue is that you need to provide the file info like extension, name, mime type, etc. The idea behind his PR is to allow plugins to programmatically upload images.

**Comment by derrickmehaffy** [comment]
> Hey @Cpaczek thanks for the PR! as this is a feature addition theres a bit of a stricter procedure on what we require for a PR to be merged.
> 
> Looking at what you've done so far:
> * it'd be great to add tests for the service function you've added
> * Can we utilise this service in the admin API and/or the content API controllers?
> 
> I'm also gonna request a review other engineers too 💪🏼 

Just a heads up Josh this is something a lot of users have asked for over the years specifically for customization (say fetching a file from a 3rd party API and uploading it in a lifecycle or cron).

It probably plays more into useful backend functions instead to replacing or extending any existing features.

**Comment by Cpaczek** [comment]
@joshuaellis I added a test, let me know if this is what you're looking for. Also, I couldn't figure out how to run it locally so I just pushed, hopefully, it passes. In the meantime could you send me instructions on how to just just 1 suite? I tried to do `yarn test:unit`  but I get a no matches error
image

**Comment by Marc-Roig** [comment]
@Cpaczek to run the test locally you can execute that same command yarn test:unit , but at the root level of the strapi project. :)  If you want only to run one specific test or tests for a file you could do

`yarn test:unit <path>/<to>/<file>/<filename>.test.js`


You can see the other scripts in the package.json there, in case you want to test the integration tests you can do `yarn test:api`.

**Comment by joshuaellis** [comment]
Closing this PR, please see this [comment](https://github.com/strapi/strapi/issues/5297#issuecomment-1430138514).

**Comment by petersg83** [comment]
Hello :)
I'm re-opening the PR so we can have a look at it from another perspective (only allowing buffer support in the backend, not allowing buffer serialization over HTTP) as there was confusion with the issue linked to this PR (#5297). This PR won't fix #5297 but addresses another need :)

**Comment by Cpaczek** [comment]
@petersg83 Does this current implementation look good? Is there anything I would need to change? If it looks good I will go ahead and finish/fix the tests.

File: packages/core/upload/server/services/upload.js [hunk 1]
```diff
@@ -180,6 +181,56 @@ module.exports = ({ strapi }) => ({
...
+      provider: config.provider,
+      folder,
+      caption,
+      alternativeText,
+      tmpWorkingDirectory: await createAndAssignTmpWorkingDirectoryToFiles({}),
+      getStream() {
+        // eslint-disable-next-line new-cap
+        return new stream.Readable.from(file);
+      },
+    };
```
Comment (nathan-pichon) [inline packages/core/upload/server/services/upload.js:229]: I think you can reuse `enhanceAndValidateFile` function and improve this one by taking into account possible stream as a file.
```javascript
async enhanceAndValidateFile(file, fileInfo = {}, metas = {}, stream = null) {
// ...
currentFile.getStream = () => stream || fs.createReadStream(file.path);
// ...
}
```

WDYT ?

**Comment by PaulBratslavsky** [comment]
This would be a great addition, ran into a use case where I also need this functionality.

**Comment by alexandrebodin** [comment]
I'll look at this during the next months to add for v5 stable

**Comment by alexandrebodin** [comment]
> [!IMPORTANT] 
> With the release of Strapi 5, we have lifted the code freeze. To ensure we can effectively review and assist with contributions, we are now only accepting PRs targeting version 5 (the develop branch).<br/> 
> As your PR targets version 4, it has been automatically closed. However, we would greatly appreciate it if you could update your contribution to align with version 5 and resubmit!

**Comment by alexandrebodin** [comment]
@cpaczek As you already have seen I have a proposed simpler version I hope to get into ~ v5.1.0

**Comment by varloc2000** [comment]
> @cpaczek As you already have seen I have a proposed simpler version I hope to get into ~ v5.1.0

Sorry, where the PR for that mentioned simpler version can be found to see it's it's merged? 
The issue is open still https://github.com/strapi/strapi/issues/15524

**Pull request closed without merging**